marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/ExecEndpointImpl.java

```diff
@@ -212,9 +263,11 @@ public Boolean call() throws InterruptedException{
                     bulkExecCallerImpl.getCallContextQueue().put(callContext);
                     submitTask(this);
                 }
-                else if(bulkExecCallerImpl.getCallContextQueue().isEmpty() &&
-                        getCallerThreadPoolExecutor().getActiveCount() <= 1) {
-                    getCallerThreadPoolExecutor().shutdown();
+                else {
```

Comment:
I'd suggest moving the decrementAndGet() call into the if() test.  That way, a different thread can't change the value before this thread gets the value, and only one thread will go to zero.  

The same change would go into the Output endpoint.

Reply:
I'm not quite sure about this one. aliveCallContextCount is AtomicInteger. It guarantees that a different thread can't change the value before this thread changes and gets the value.

Reply:
Thanks for the explanation. Changed the statement as below. 

if (aliveCallContextCount.decrementAndGet() == 0) {
                        getCallerThreadPoolExecutor().shutdown();
                    }